xmppserver/src/main/java/org/jivesoftware/openfire/IQRouter.java

```diff
@@ -296,18 +303,27 @@ private void handle(IQ packet) {
         }
         if (packet.getID() != null && (IQ.Type.result == packet.getType() || IQ.Type.error == packet.getType())) {
             // The server got an answer to an IQ packet that was sent from the server
+
+            // If there's a listener for this result at all, then it's likely that that listener had been registered
+            // on this cluster node. For efficiency, try the local cluster node before triggering tasks in the rest
+            // of the cluster.
             IQResultListener iqResultListener = resultListeners.remove(packet.getID());
             if (iqResultListener != null) {
```

Comment:
Appreciate you didn’t change this, but would be very attempted to use the return value of the remove() method to obtain the resulListener to reduce potential concurrency issues.

Reply:
Other than the trivial comment above, this all looks good to me. I don’t recall anything like this, but it seems to make sense.

Reply:
I'm not sure what you mean.

Reply:
Yeah, misread something - ignore that comment!